### ISSUE TITLE
PageView telemetry missing duration property when autoTrackPageVisitTime = true

### ISSUE THREAD

**Issue by josundt** [Microsoft/ApplicationInsights-JS#835]
The error I reported in #826, #827 seems to be fixed in 2.0.0-rc.2, but still the **duration** property is missing on the pageView telemetry items in Application Insights.

This property used to be present on pageView telemetry before we tested upgrading to JS SDK 2.0.0-x, and I expected this property to be present when I have configured the TelemetryClient with **autoTrackPageVisitTime = true**

_PS! I also tried setting the **overridePageViewDuration** setting to **true** (even though I am not sure what that implies). It anyway did not make any difference._

We are simply calling the trackPageView method from our SPA, passing name, url, with some custom properties and measurements.

**Comment by markwolff** [comment]
Hey @josundt you will need to add the `duration` field to `properties` in `trackPageView`. I'll add a note of this in the docs.
```js
appInsights.trackPageView({
  name: 'pageview with duration = 00:05:00.000',
  properties: { duration: 300000 }
});
```
`overridePageViewDuration` just changes the automatic calculation of the duration field to be the interval between pageLoad and your call to trackPageView. Otherwise, the default behavior is to use the navigator timing API.

**Comment by josundt** [comment]
In our code, we make a call to trackPageView when the user enters a (SPA) page. At this point we obviously do not know how long the user will stay on the page :-).

I thought the JS SDK automatically updated the duration property of the previously tracked pageView telemetry item when trackPageView is called for the second time or more - within the same user/browser session.

Am I wrong?

**Comment by markwolff** [comment]
`autoTrackPageVisitTime` tracks the view time of the previous instrumented page for each call to trackPageView, and sends it as a Metric.

If you set `overridePageViewDuration: true`, this only changes the `duration` field of your subsequent `trackPageView` calls to be the time since the page loaded, although the page load time may not be correct for SPA pages. So if you call trackPageView every second, the duration will be increased by 1 second on each call.

Are you just trying to match functionality with 1.0.20 of the SDK?

**Comment by josundt** [comment]
I work for a company that makes a web software suite consisting of multiple web applications, mainly SPA+AspNetCore+NetCore background services (hosted in AKS). 

We are trying to establish a standard for the basic contextual (user/device/environment) standard/custom properties/metrics to be set for AI telemetry across all applications across all telemetry sources (browser/webserver/background-service etc).

To know that I have configured the new 2.x SDK properly, I compare the new 2.x telemetry to the old 1.x to verify that properties are initialized with corresponding values. 

I just noticed that the duration property from pageView telemetry was lost after upgrading and wondered why, since there seems to be some built-in features in the SDK to track page view durations, and this seems to have worked out-of-the box in 1.x

I can't find much documentation on this. so would you please be kind and explain how I am supposed to get the duration tracked? As I said earlier, all we are doing from code is calling the trackPageView method when a SPA page navigation occurs with the url/name of the page the user is entering. Do we need to call it again when the user leaves the page?

BTW: I also noticed that the client_Type property was no longer automatically set to "Browser", but I found a way to add a tag to ensure that that property gets set like before. Not sure why this is not default behavior anymore.

**Comment by markwolff** [comment]
Thanks for adding some clarification. I just released 2.0.0-rc3 which fixes the missing duration field. In general, the duration behavior should not have changed at all from 1.x to 2.x, and so the stuff you had working before should behave the same here.

The duration field in trackPageView is defined as the time it took for the page to load (the pageload event is triggered in the user's browser). I'm not sure if the duration you are seeking is how long a user viewed the page. If so, `autoTrackPageVisitTime` accomplishes this, but sends it as a customMetric to the Azure Portal. This is how this was accomplished in 1.x as well.

There is some info on all of this here:
https://docs.microsoft.com/en-us/azure/azure-monitor/app/javascript#page-view-properties

**Comment by josundt** [comment]
I had a look at the link you provided and did some more digging. I eventually found that the **duration** property is intended for measuring the time it takes to **load** a page (_not_ the time a user **stays on a page** like I believed earlier).

Page load time measurement is definitely as relevant in SPAs as in traditional web applications. The perceived latency when loading a new view/page is what this duration property should reflect. To the user the latency experience is more or less the same regardless of what technology that's being used.

The technical definition of "page load" might be quite different though:

**Page load in a traditional web application:**
* Loading an HTML document with all assets into the browser window.

**Page load time in a typical SPA framework:**
* JS "router" detects a browser address URL change; then intercepts and maps the new URL to a new "view" through a routing table.
* Teardown of the view being left (deactivation/disposal of JS component, detaching DOM elements).
* Buildup of the view being entered (initialization/activation of JS component, potentially fetching data from server APIs asynchronously).
* Data binding/rendering of the new view (Creating/attaching DOM elements etc).

From the link you provided in your previous comment, where three different possible behaviors for duration measurement was mentioned, only the third of these options really fit a SPA application (measuring and sending **duration** as argument to trackPageView). 

It is highly possible to measure the time spent for the SPA "page load" steps, and I can provide the duration property. So I tested this option **_but found one more bug_**:

* I set **overridePageViewDuration** to true
* I call 
```typescript
trackPageView({ name, url  }, { duration: timeSpentInMilliseconds })
```

But I found that the trackPageView method in the Initialization class (@microsoft/applicationinsights-web/src/Initiailization.ts:85) does not forward the second argument to the corresponding method in the ApplicationInsights class.

Current:
(@microsoft/applicationinsights-web/src/Initiailization.ts:85)
```typescript
    public trackPageView(pageView?: IPageViewTelemetry) {
        const inPv = pageView || {};
        this.appInsights.trackPageView(inPv);
    }
```
Expected:
```typescript
    public trackPageView(pageView?: IPageViewTelemetry, customProperties?: ICustomProperties) {
        const inPv = pageView || {};
        this.appInsights.trackPageView(inPv, customProperties);
    }
```

Can you please fix - everything works

**Comment by markwolff** [comment]
You should use the `properties` field of `IPageViewTelemetry`. the `customProperties` field is deprecated and is not exposed in our API.
https://github.com/Microsoft/ApplicationInsights-JS/blob/d9a2f3cdb23b7f093e1f9bc88638e0fcd375aed4/vNext/AISKU/src/Initialization.ts#L80-L85

```
trackPageView({ name: "thename", url: "/theurl/", properties: { duration: timeSpentInMilliseconds } });
```

**Comment by josundt** [comment]
I tried that - that does not work. The duration property was not set in the telemetry in Application Insights when putting duration in the "properties" object of IPageViewTelemetry.

I also believed that the "properties" object in IPageViewTelemetry needed to be a shallow object with string values only `{ [key: string]: string },` while "measurements" is a shallow object with numeric values only `{ [key: string]: number }`.

When testing the deprecated customProperties argument it worked.

**Comment by markwolff** [comment]
The requirement for properties to be a string, string map was removed in #819. Properties can now contain objects or strings, etc.

When you specify a duration in the properties field, we remove it from properties and set it in your actual telemetry item instead (in `baseData`).

```js
appInsights.trackPageView({ name: 'pageview with duration = 00:05:00.000', properties: { duration: 300000 } });

```

![image](https://user-images.githubusercontent.com/5515583/56247957-553f1a00-605b-11e9-80d2-34b1982d3b1b.png)

**Comment by josundt** [comment]
Thanks, you're right. Passing **IPageViewTelemetry.properties.duration** as number (milliseconds) sets the duration property correctly. I was not aware that "properties" started supporting numeric values and could for some reason not make this work last time I tried.

PS! My current setup with `overridePageViewDuration = true` and measuring/passing the duration on calls to `trackPageView` is probably the setup wanted by most Single Page Applications.

The documentation in the link you sent me a couple of comments back could definitely be more clear about this - it could mention that the third option is probably what you want if you're working with a Single Page Application, and it could be more specific about how to pass the duration property when measuring it yourself.

I also found the custom metric with the PageVisitTimes, so now everything is finally clear to me!

Thanks for all your help! :-)
